### PR TITLE
[SPARK-35271][ML][PYSPARK] Fix: After CrossValidator/TrainValidationSplit fit raised error, some backgroud threads may still continue run or launch new spark jobs

### PR DESCRIPTION
**Pull request by WeichenXu123**
### What changes were proposed in this pull request?

Fix: After CrossValidator/TrainValidationSplit fit raised error, some backgroud threads may still continue run or launch new spark jobs

### Why are the changes needed?

Bugfix

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit tests.

### CODE REVIEW

**Comment by srowen** [comment]
What is the Spark JIRA? that looks like something internal

**Review by srowen — COMMENTED** [review]
The change looks OK

File: python/pyspark/util.py [hunk 1]
```diff
@@ -263,6 +264,46 @@ def _parse_memory(s):
...
+def _on_inheritable_thread_start():
+    from pyspark import SparkContext
+    sc = SparkContext._active_spark_context
+    if isinstance(sc._gateway, ClientServer):
+        # Here's when the pinned-thread mode (PYSPARK_PIN_THREAD) is on.
+        properties = sc._jsc.sc().getLocalProperties().clone()
+        sc._jsc.sc().setLocalProperties(properties)
+
+
+def _on_inheritable_thread_stop():
```
Comment (HyukjinKwon) [inline python/pyspark/util.py:276]: Can we dedup the codes with `InheritableThread`?

File: python/pyspark/util.py [hunk 2]
```diff
@@ -263,6 +264,46 @@ def _parse_memory(s):
...
+                    if connections[i] is thread_connection:
+                        connections[i].close()
+                        del connections[i]
+                        break
+                else:
+                    # Just in case the connection was not closed but removed from the queue.
+                    thread_connection.close()
+
+
+def inheritable_thread(f):
```
Comment (HyukjinKwon) [inline python/pyspark/util.py:296]: Should also add some docstrings like `InheritableThread`

File: mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala [hunk 3]
```diff
@@ -183,7 +190,16 @@ class CrossValidator @Since("1.2.0") (@Since("1.4.0") override val uid: String)
...
 
       // Wait for metrics to be calculated
-      val foldMetrics = foldMetricFutures.map(ThreadUtils.awaitResult(_, Duration.Inf))
+      val foldMetrics = try {
+        foldMetricFutures.map(ThreadUtils.awaitResult(_, Duration.Inf))
+      } finally {
+        subTaskFailed = true
+        Thread.sleep(1000)
+        val sparkContext = sparkSession.sparkContext
+        sparkContext.cancelJobGroup(
```
Comment (HyukjinKwon) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:199]: Should we only cancel on failure?

File: mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala [hunk 4]
```diff
@@ -168,9 +168,16 @@ class CrossValidator @Since("1.2.0") (@Since("1.4.0") override val uid: String)
       val validationDataset = sparkSession.createDataFrame(validation, schemaWithoutFold).cache()
       instr.logDebug(s"Train split $splitIndex with multiple sets of parameters.")
 
+      var subTaskFailed = false
```
Comment (HyukjinKwon) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:171]: I think it should be either atomic boolean or `@volatile`?

File: python/pyspark/util.py (hunk 2 above)
Comment (HyukjinKwon) [reply python/pyspark/util.py:296]: I think it should also show some example about how to use

Comment (HyukjinKwon) [inline python/pyspark/util.py:296]: I would name it a bit differently like `inheritable_thread_target` (given that Python's thread takes `target`) or `inheritable_thread_func` because this function itself isn't a thread.

**Comment by HyukjinKwon** [comment]
The idea seems good!

**Comment by srowen** [comment]
Really minor whitespace thing:
`[error] /home/jenkins/workspace/SparkPullRequestBuilder/mllib/src/test/scala/org/apache/spark/ml/tuning/CrossValidatorSuite.scala:583:0: Whitespace at end of line
`

File: python/pyspark/util.py [hunk 5]
```diff
@@ -263,6 +264,69 @@ def _parse_memory(s):
...
+
+    Also, note that pinned thread mode does not close the connection from Python
+    to JVM when the thread is finished in the Python side. With this wrapper, Python
+    garbage-collects the Python thread instance and also closes the connection
+    which finishes JVM thread correctly.
+
+    When the pinned thread mode is off, it return the original ``f``.
+    :param f: the original thread target.
+
+    .. versionadded:: 3.1.0
```
Comment (HyukjinKwon) [inline python/pyspark/util.py:290]: I think you meant: 3.2.0

File: mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala [hunk 6]
```diff
@@ -161,11 +169,26 @@ class TrainValidationSplit @Since("1.5.0") (@Since("1.5.0") override val uid: St
...
+        throw e
+    }
+    finally {
+      if (subTaskFailed) {
+        Thread.sleep(1000)
+        val sparkContext = dataset.sparkSession.sparkContext
+        sparkContext.cancelJobGroup(
+          sparkContext.getLocalProperty(SparkContext.SPARK_JOB_GROUP_ID)
+        )
+      }
```
Comment (HyukjinKwon) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala:187]: i think we could do this as below: without `subTaskFailed` .. ?

```scala
    catch {
      case e: Throwable =>
        val sparkContext = dataset.sparkSession.sparkContext
        sparkContext.cancelJobGroup(
          sparkContext.getLocalProperty(SparkContext.SPARK_JOB_GROUP_ID)
        )
```

File: mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala [hunk 7]
```diff
@@ -161,11 +169,26 @@ class TrainValidationSplit @Since("1.5.0") (@Since("1.5.0") override val uid: St
...
+      metricFutures.map(ThreadUtils.awaitResult(_, Duration.Inf))
+    }
+    catch {
+      case e: Throwable =>
+        subTaskFailed = true
+        throw e
+    }
+    finally {
+      if (subTaskFailed) {
+        Thread.sleep(1000)
```
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala:182]: This sleep is for:

each trial task which thread already running, may took some time running before it launch spark job, if here we cancel job immediately, then we may miss killing the spark job which will be spawned soon

File: python/pyspark/ml/tuning.py [hunk 8]
```diff
@@ -730,13 +733,40 @@ def _fit(self, dataset):
...
+                    if is_pinned_thread_mode():
+                        try:
+                            time.sleep(1)
+                            sc = dataset._sc
+                            sc.cancelJobGroup(sc.getLocalProperty("spark.jobGroup.id"))
+                        except:
+                            pass
+                    else:
+                        warnings.warn("CrossValidator {} fit call failed but some spark jobs "
+                                      "may still running for unfinished trials. Enable pyspark "
```
Comment (HyukjinKwon) [inline python/pyspark/ml/tuning.py:764]: Hm, why is it inconsistent with `TrainValidationSplit`? Seems like `TrainValidationSplit` will always cancel but here only cancel when pinned thread mode is on.

File: mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala (hunk 7 above)
Comment (HyukjinKwon) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala:182]: Would you mind elabourating why we should sleep here? I should avoid relying on sleep in the main codes or test codes whenever possible.

**Review by HyukjinKwon — COMMENTED** [review]
cc @viirya too

File: dev/sparktestsupport/modules.py [hunk 9]
```diff
@@ -565,6 +565,7 @@ def __hash__(self):
         "pyspark.ml.tests.test_stat",
         "pyspark.ml.tests.test_training_summary",
         "pyspark.ml.tests.test_tuning",
+        "pyspark.ml.tests.pyspark.ml.tests.test_tuning_on_pin_thread_mode",
```
Comment (viirya) [inline dev/sparktestsupport/modules.py:568]: Does `pyspark.ml.tests` duplicate?

File: python/pyspark/util.py (hunk 5 above)
Comment (WeichenXu123) [reply python/pyspark/util.py:290]: i think we need backport to 3.1

File: mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala [hunk 10]
```diff
@@ -183,11 +190,28 @@ class CrossValidator @Since("1.2.0") (@Since("1.4.0") override val uid: String)
...
-      trainingDataset.unpersist()
-      validationDataset.unpersist()
+      val foldMetrics = try {
+        foldMetricFutures.map(ThreadUtils.awaitResult(_, Duration.Inf))
+      }
+      catch {
+        case e: Throwable =>
+          subTaskFailed = true
+          try {
+            Thread.sleep(1000)
```
Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:200]: Why we need sleep 1000?

Comment (WeichenXu123) [reply mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:200]: https://github.com/apache/spark/pull/32399#discussion_r629087709

File: .idea/vcs.xml [hunk 11]
```diff
@@ -1,36 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (srowen) [inline .idea/vcs.xml:1]: Is it intended to remove this file? looks unrelated.

File: mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala [hunk 12]
```diff
@@ -168,9 +169,19 @@ class CrossValidator @Since("1.2.0") (@Since("1.4.0") override val uid: String)
...
+      val sparkContext = sparkSession.sparkContext
+      val oldJobGroup = sparkContext.getLocalProperty(SparkContext.SPARK_JOB_GROUP_ID)
+      val cvJobGroup = s"${this.uid}_job_group"
+      sparkContext.setLocalProperty(SparkContext.SPARK_JOB_GROUP_ID, cvJobGroup)
+      @volatile var subTaskFailed = false
       // Fit models in a Future for training in parallel
       val foldMetricFutures = epm.zipWithIndex.map { case (paramMap, paramIndex) =>
         Future[Double] {
+          if (subTaskFailed) {
+            throw new RuntimeException(
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:181]: The exception doesn't matter here? so RuntimeException is OK? otherwise that's probably overly broad. At least, IllegalStateException or something. But if it's just going to kill the Future, OK

File: mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala [hunk 13]
```diff
@@ -183,11 +194,27 @@ class CrossValidator @Since("1.2.0") (@Since("1.4.0") override val uid: String)
...
+        case e: Throwable =>
+          subTaskFailed = true
+          try {
+            Thread.sleep(1000)
+            sparkContext.cancelJobGroup(cvJobGroup)
+          } catch {
+            case _: Throwable => ()
+          }
+          throw e
+      }
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:210]: Nit: finally on this line

File: mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala [hunk 14]
```diff
@@ -161,11 +174,28 @@ class TrainValidationSplit @Since("1.5.0") (@Since("1.5.0") override val uid: St
...
+    }
+    catch {
+      case e: Throwable =>
+        subTaskFailed = true
+        try {
+          Thread.sleep(1000)
+          val sparkContext = sparkSession.sparkContext
+          sparkContext.cancelJobGroup(tvsJobGroup)
+        } catch {
+          case _: Throwable => ()
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala:188]: Nit: is () necessary versus just nothing here?

File: python/pyspark/util.py (hunk 5 above)
Comment (srowen) [reply python/pyspark/util.py:290]: I'd also not backport unless it's essential - how bad is it?

File: mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala (hunk 14 above)
Comment (WeichenXu123) [reply mllib/src/main/scala/org/apache/spark/ml/tuning/TrainValidationSplit.scala:188]: I think this is a standard way representing return "Unit", https://stackoverflow.com/questions/53415664/how-do-i-explicitly-return-unit

File: python/pyspark/util.py (hunk 5 above)
Comment (WeichenXu123) [reply python/pyspark/util.py:290]: Ok we don't need backport it in Apache/spark.

File: mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala (hunk 12 above)
Comment (WeichenXu123) [reply mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:181]: Yes just for kill the future thread.
I already add message in the runtimeException. So I think it is fine.
`IllegalStateException` means "Signals that a method has been invoked at an illegal or inappropriate time" so it may be not a proper exception class name.

**Comment by HyukjinKwon** [comment]
re: https://github.com/apache/spark/pull/32399#discussion_r629087709

How does it guarantee that it kill the Spark jobs? After one second, we might be on the step 3, and it ends up with not killing any job.

File: mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala [hunk 15]
```diff
@@ -168,9 +169,19 @@ class CrossValidator @Since("1.2.0") (@Since("1.4.0") override val uid: String)
       val validationDataset = sparkSession.createDataFrame(validation, schemaWithoutFold).cache()
       instr.logDebug(s"Train split $splitIndex with multiple sets of parameters.")
 
+      val sparkContext = sparkSession.sparkContext
+      val oldJobGroup = sparkContext.getLocalProperty(SparkContext.SPARK_JOB_GROUP_ID)
+      val cvJobGroup = s"${this.uid}_job_group"
+      sparkContext.setLocalProperty(SparkContext.SPARK_JOB_GROUP_ID, cvJobGroup)
```
Comment (HyukjinKwon) [inline mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:175]: Hm, wouldn't this unset previous job group. If users want to cancel `oldJobGroup` group, looks like it's not going to work

**Comment by WeichenXu123** [comment]
> re: [#32399 (comment)](https://github.com/apache/spark/pull/32399#discussion_r629087709)
> 
> How does it guarantee that it kill the Spark jobs? After one second, we might be on the step 3, and it ends up with not killing any job.

no guarantee…:joy:

But the assumption is:
def trial_thread_target():
   if subTaskFailed:
       raise Error()
   else:
      # 1. run some code here
      # 2. launch a spark job...
      # 3. run some code here
      # 4. launch a second spark job...
      # ....
the 1/3/.. step took little time because most of training workloads will be run via spark job.

File: mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala (hunk 15 above)
Comment (WeichenXu123) [reply mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala:175]: Yes this is a corner case. I don't have good idea.
If we support cancelling by job group prefix, then it could address this issue.
CC @srowen Any thoughts ?

**Comment by WeichenXu123** [comment]
Due to flakiness in the fix, close this for now, I will reopen this when finding better solution.
